{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2.2. (Mouse) Differential expression analysis\n",
    "\n",
    "We extracted differentially expressed genes for the 3 time points compared to the baseline (Day 0) and in pairwise-manner (Day T vs T-1).\n",
    "\n",
    "Differential expression analysis was done with the robust QLF test and the adjusted p-value was independently computed for each comparison.\n",
    "\n",
    "## Input\n",
    "\n",
    "* `data-create_networks/mouse/mouse_DRG_WT_timecourse_counts.txt`: raw counts generated by featureCounts from the mouse dataset.\n",
    "\n",
    "## Output\n",
    "\n",
    "* `data-create_networks/mouse/dea_quality.pdf`: quality control of the differential expression analysis.\n",
    "* Files within the folder `data-create_networks/mouse/comparisons`: results of the DEA for the comparisons.\n",
    "* `data-create_networks/mouse/mouse_logCPM.txt`: normalized counts in log CPM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: limma\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"edgeR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_file = '../../../data-create_networks/mouse/mouse_DRG_WT_timecourse_counts.txt'\n",
    "dea_quality_file = '../../../data-create_networks/mouse/dea_quality.pdf'\n",
    "results_folder = '../../../data-create_networks/mouse/comparisons/'\n",
    "logCPM_file = '../../../data-create_networks/mouse/mouse_logCPM.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data\n",
    "counts = read.table( count_file, skip = 1, header = TRUE, row.names = 1 )\n",
    "counts = counts[ , 6:ncol(counts) ] # remove first columns\n",
    "\n",
    "# rename header\n",
    "cols = gsub( 'Aligned.out.bam', '', colnames(counts) )\n",
    "cols = gsub( 'bam_trimmed.DRG_WT_', '', cols )\n",
    "colnames(counts) = paste('Day', cols, sep='')\n",
    "\n",
    "# experiment design\n",
    "N_replicates = dim(counts)[2]/2\n",
    "group = factor( rep( 1:N_replicates, each=2 ), levels=1:N_replicates )\n",
    "design = model.matrix( ~group )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes before filtering: 24421 \n",
      "Number of genes after filtering: 17328 \n",
      "BCV = 0.08850174"
     ]
    }
   ],
   "source": [
    "initCds = DGEList( counts, group = group )\n",
    "\n",
    "# filter lowly expressed genes\n",
    "keepGenes = filterByExpr( initCds )\n",
    "initCds = initCds[ keepGenes, , keep.lib.sizes=FALSE ]\n",
    "\n",
    "cat( \"Number of genes before filtering:\", nrow( counts ), \"\\n\" )\n",
    "cat( \"Number of genes after filtering:\", nrow( initCds$counts ), \"\\n\" )\n",
    "\n",
    "# compute model\n",
    "initCds = calcNormFactors( initCds )\n",
    "initCds = estimateDisp( initCds, design, robust = TRUE )\n",
    "fit = glmQLFit( initCds, design, robust = TRUE )\n",
    "cat( \"BCV =\", sqrt( initCds$common.dispersion ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(dea_quality_file)\n",
    "\n",
    "plotMDS( initCds )\n",
    "plotBCV( initCds )\n",
    "\n",
    "qlfTests = list()\n",
    "contrasts = list( c(0,1,0,0), c(0,0,1,0), c(0,0,0,1), c(0,-1,1,0), c(0,0,-1,1) )\n",
    "names = c('Day1vs0', 'Day3vs0', 'Day30vs0', 'Day3vs1', 'Day30vs3')\n",
    "for( i in 1:5 ){\n",
    "    name = names[i]\n",
    "    # test\n",
    "    qlfTests[name] = list( glmQLFTest( fit, contrast = contrasts[[i]] ) )\n",
    "    # adjusted p-value\n",
    "    padj = p.adjust( qlfTests[[name]]$table$PValue, method = \"BH\" )\n",
    "    qlfTests[[name]]$table$FDR = padj\n",
    "    # display\n",
    "    plotQLDisp( qlfTests[[name]], main=paste(name, \"glm edgeR QL Dispersion\") )\n",
    "    hist(qlfTests[[name]]$table$PValue, 150,\n",
    "         main=paste(name, \"glm edgeR tagwise dispersion\"),xlab=\"raw pvalue\")\n",
    "}\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### exports\n",
    "i = 1\n",
    "for( name in names(qlfTests) ){\n",
    "    write.table( qlfTests[[name]]$table,\n",
    "                 file=paste( results_folder, formatC(i, width=2, flag=\"0\"), \"_mouse_\",\n",
    "                             name, \".txt\", sep='' ),\n",
    "                 sep = \"\\t\", quote=FALSE, row.names=TRUE, col.names=NA )\n",
    "    i = i + 1\n",
    "}\n",
    "\n",
    "write.table( cpm(initCds, prior.count=2, log=TRUE),\n",
    "             file = logCPM_file,\n",
    "             sep = \"\\t\", quote=FALSE, row.names=TRUE, col.names=NA )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:timenexus_env]",
   "language": "R",
   "name": "conda-env-timenexus_env-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
